llama_index/vector_stores/redis.py

```diff
@@ -0,0 +1,323 @@
+
+        try:
+            results = self._redis_client.ft(self._index_name).search(
+                redis_query, query_params=query_params
+            )
+        except RedisResponseError as e:
+            _logger.error(f"Error querying index {self._index_name}: {e}")
+            raise e
+
+        ids = []
+        nodes = []
+        scores = []
+        for doc in results.docs:
+            node = Node(
+                text=doc.text,
+                doc_id=doc.id,
+                embedding=None,
+                relationships={DocumentRelationship.SOURCE: doc.doc_id},
+            )
+            ids.append(doc.id)
+            nodes.append(node)
+            scores.append(1 - float(doc.vector_score))
+        _logger.info(f"Found {len(nodes)} results for query with id {ids}")
+
+        return VectorStoreQueryResult(nodes=nodes, ids=ids, similarities=scores)
+
+    def persist(self, in_background=True) -> None:
+        """Persist the vector store to disk.
+
+        Args:
+            in_background (bool, optional): Persist in background. Defaults to True.
+        """
+        if in_background:
+            _logger.info("Saving index to disk in background")
+            self._redis_client.bgsave()
+        else:
+            _logger.info("Saving index to disk")
+            self._redis_client.save()
+
+    def _create_index(self) -> None:
+        # should never be called outside class and hence should not raise importerror
+        from redis.commands.search.field import TagField, TextField
+        from redis.commands.search.indexDefinition import IndexDefinition, IndexType
+
+        # Create Index
+        default_fields = [
+            TextField("text", weight=1.0),
+            TagField("doc_id", sortable=False),
+            TagField("id", sortable=False),
+        ]
+        # add vector field to list of index fields. Create lazily to allow user
+        # to specify index and search attributes in creation.
+        fields = default_fields + [
+            self._create_vector_field(self._vector_field, **self._index_args)
+        ]
+
+        _logger.info(f"Creating index {self._index_name}")
+        self._redis_client.ft(self._index_name).create_index(
+            fields=fields,
+            definition=IndexDefinition(
+                prefix=[self._prefix], index_type=IndexType.HASH
+            ),  # TODO support JSON
+        )
+
+    def _index_exists(self) -> bool:
+        # use FT._LIST to check if index exists
+        indices = convert_bytes(self._redis_client.execute_command("FT._LIST"))
+        return self._index_name in indices
+
+    def _prep_query(self, top_k: int = 10) -> str:
+        # returns vector_score by default
+        return_fields = ["id", "doc_id", "text", self._vector_key, "vector_score"]
+
+        redis_query = get_redis_query(
+            return_fields=return_fields, top_k=top_k, vector_field=self._vector_field
+        )
+        return redis_query
+
+    def _create_vector_field(
```

Comment:
It wasn't immediately obvious to me that users can pass in index args to configure this. Is this documented somewhere else? Do you expect the users to change these?

Reply:
In most cases, no. But advanced users may want to, hence the approach. Added pointers to the docs. They are named the exact same so it's easy to recognize for people who know what they are doing. For others, it'll be a footgun.

Reply:
great